.github/CONTRIBUTING.md

```diff
@@ -1,8 +1,65 @@
+
+**Codelabs** are interactive tutorials that are written in markdown syntax and published at [blocklycodelabs.dev](https://blocklycodelabs.dev). Codelabs mix natural language, code samples, and screenshots. The target user is following along and running the code as they read.
+
+The codelabs directory has a [template](https://github.com/google/blockly-samples/blob/master/codelabs/template.md) and one folder per codelab. Each codelab's folder contains the codelab (as a markdown file), and all associated assets (pngs, gif, etc).
+
+**Examples** are sample projects showing how to use Blockly features. Examples contain mostly code and a demo web page. Example code should be extremely well-commented to make it easy to copy. The target user may be reading the code, running it locally, or copying code snippets.
+
+The examples directory has one folder per example. Each example can be run with `npm install && npm run start`, and has a `README.md` file with additional context or instructions.
+
+**Plugins** are self-contained pieces of code that add functionality to Blockly. Plugins can add fields, define themes, create renderers, and much more. The target user is a developer who finds and uses the plugin through npm. Plugins defined in this repository are *first-party* plugins, which means that they are supported by the Blockly team.
+
+The plugins directory has one folder per plugin. To create the correct directory structure for your plugin, use [@blockly/create-package](https://github.com/google/blockly-samples/tree/master/plugins/dev-create#blocklycreate-package-).
+
+## How can I contribute?
+
+- Suggest plugins, examples, or codelabs
+- Implement plugins, examples, or codelabs
+- Report bugs
+- Fix bugs
+- Reproduce bugs
+
+## How do I add a plugin?
+
+Plugins go through four stages: suggestion, discussion, implementation, and publishing:
+
+A plugin starts as a **suggestion**. You can suggest a plugin by creating a new issue with the Feature Request template. To write a good suggestion:
+- Describe the desired functionality.
+- Describe the API the plugin would expose.
+- List alternatives you've considered.
+- List any APIs that need to be added or changed in core Blockly to support the plugin.
+- Include screenshots, GIFs, or mock-ups if the plugin includes UI features.
+
+Before making a suggestion, [search](https://github.com/google/blockly-samples/issues?q=is%3Aopen+is%3Aissue+label%3A%22category%3A+plugin%22+label%3A%22type%3A+feature+request%22) for a related issue. If you find one, add a comment to the existing issue instead of opening a new one.
+
+Next a plugin will go into the **discussion** phase. This phase includes:
+- Clarification of the desired functionality.
+- Clarification of the plugin's API.
+- Planning for implementation.
+- Planning for tests.
+- Discussion of API changes in core Blockly.
+- Breaking large plugins into implementation steps.
+- Plugin naming, based on naming conventions that we will publish Real Soon Now.
```

Comment:
I think these are all really great points. Once you have all of this stuff figured out you can basically blaze right through the coding. But I wonder if the process for /how/ this stuff gets figured out could be a bit more clear.

For example take the "Planning for implementation" bullet point: Who is in charge of planning the implementation? Is it the one who suggested the plugin? What if they're just suggesting the plugin and don't want to implement it? What does an implementation plan look like? Does it go through it's own review process?

And what do you want to end up with at the end of this discussion phase? A series of new issues? A list of bullet points that gets attached to the original issue? Or something else?

How do you know when the discussion phase is complete?

I think it's also important to note that this is a place where you can easily add too much bureaucracy. By which I mean: any extra steps that don't make the process easier for contributors. It's just a good thing to keep in mind that we probably want to keep this process small for now, we can always add more steps later if they become absolutely necessary.

Reply:
Lots of good questions here, which really reinforces the idea that this needs to be its own page in the wiki.